Sources/_MatchingEngine/Regex/AST/MatchingOptions.swift

```diff
@@ -0,0 +1,82 @@
+extension AST {
+  /// An option written in source that changes matching semantics.
+  public struct MatchingOption: Hashable {
+    public enum Kind {
+      // PCRE options
+      case caseInsensitive          // i
+      case allowDuplicateGroupNames // J
+      case multiline                // m
+      case noAutoCapture            // n
+      case singleLine               // s
+      case reluctantByDefault       // U
+      case extended                 // x
+      case extraExtended            // xx
+
+      // ICU options
+      case unicodeWordBoundaries    // w
+
+      // Oniguruma options
+      case asciiOnlyDigit           // D
+      case asciiOnlyPOSIXProps      // P
+      case asciiOnlySpace           // S
+      case asciiOnlyWord            // W
+
```

Comment:
@natecook1000 Should we include the `X`, `u` & `b` options from https://github.com/apple/swift-experimental-string-processing/pull/64 here?